src/nupic/experimental/ExtendedTemporalMemory.cpp

```diff
@@ -0,0 +1,1316 @@
+      : std::binary_search(prevActiveExternalCells.begin(),
+                           prevActiveExternalCells.end(),
+                           synapseData.presynapticCell - connections.numCells());
+
+    Permanence permanence = synapseData.permanence;
+
+    if (isActive)
+    {
+      permanence += permanenceIncrement;
+    }
+    else
+    {
+      permanence -= permanenceDecrement;
+    }
+
+    permanence = min(permanence, (Permanence)1.0);
+    permanence = max(permanence, (Permanence)0.0);
+
+    if (permanence < EPSILON)
+    {
+      connections.destroySynapse(synapse);
+    }
+    else
+    {
+      connections.updateSynapsePermanence(synapse, permanence);
+    }
+  }
+
+  if (connections.numSynapses(segment) == 0)
+  {
+    connections.destroySegment(segment);
+  }
+}
+
+static void growSynapses(
+  Connections& connections,
+  Random& rng,
+  Segment segment,
+  UInt32 nDesiredNewSynapses,
+  const vector<CellIdx>& internalCandidates,
+  const vector<CellIdx>& externalCandidates,
+  Permanence initialPermanence,
+  bool formInternalConnections)
+{
+  vector<CellIdx> candidates;
+
+  if (formInternalConnections)
+  {
+    candidates.reserve(internalCandidates.size() + externalCandidates.size());
+    candidates.insert(candidates.begin(), internalCandidates.begin(),
+                      internalCandidates.end());
+  }
+  else
+  {
+    candidates.reserve(externalCandidates.size());
+  }
+
+  for (CellIdx cell : externalCandidates)
+  {
+    candidates.push_back(cell + connections.numCells());
+  }
+
+  // Instead of erasing candidates, swap them to the end, and remember where the
+  // "eligible" candidates end.
+  auto eligibleEnd = candidates.end();
+
+  // Remove cells that are already synapsed on by this segment
+  for (Synapse synapse : connections.synapsesForSegment(segment))
+  {
+    CellIdx presynapticCell =
+      connections.dataForSynapse(synapse).presynapticCell;
+    auto ineligible = find(candidates.begin(), eligibleEnd, presynapticCell);
+    if (ineligible != eligibleEnd)
+    {
+      eligibleEnd--;
+      std::iter_swap(ineligible, eligibleEnd);
+    }
+  }
+
+  const UInt32 nActual =
+    std::min(nDesiredNewSynapses,
+             (UInt32)std::distance(candidates.begin(), eligibleEnd));
+
+  // Pick nActual cells randomly.
+  for (UInt32 c = 0; c < nActual; c++)
+  {
+    size_t i = rng.getUInt32(std::distance(candidates.begin(), eligibleEnd));;
+    connections.createSynapse(segment, candidates[i], initialPermanence);
+    eligibleEnd--;
+    std::swap(candidates[i], *eligibleEnd);
+  }
+}
+
+static void activatePredictedColumn(
+  vector<CellIdx>& activeCells,
+  vector<CellIdx>& winnerCells,
+  Connections& connections,
+  const ExcitedColumnData& excitedColumn,
+  bool learn,
+  const vector<CellIdx>& prevActiveInternalCells,
+  const vector<CellIdx>& prevActiveExternalCells,
+  Permanence permanenceIncrement,
+  Permanence permanenceDecrement)
+{
+  auto active = excitedColumn.activeSegmentsBegin;
+  do
```

Comment:
It has to do everything in the block at least once. So I think a do-while is appropriate -- it puts the condition at the end. (Note that this is copied directly from TemporalMemory)